File: src/codeforces/set0/set1/set100/set101/e/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	_, _, _, best, res := drive(reader)
	fmt.Println(best)
	fmt.Println(res)
}

func drive(reader *bufio.Reader) (x []int, y []int, p int, best int, res string) {
	var n, m int
	fmt.Fscan(reader, &n, &m, &p)
	x = make([]int, n)
	for i := range n {
		fmt.Fscan(reader, &x[i])
	}
	y = make([]int, m)
	for i := range m {
		fmt.Fscan(reader, &y[i])
	}
	best, res = solve(x, y, p)
	return
}

var (
	gx, gy       []int
	gp, gn, gm   int
	garr1, garr2 []int
	garr3        []int
	gpath        []byte
)

func solve(x []int, y []int, p int) (best int, res string) {
	gn, gm, gp = len(x), len(y), p
	gx, gy = x, y
	garr1 = make([]int, gn)
	garr2 = make([]int, gn)
	garr3 = make([]int, gn)
	gpath = make([]byte, gn+gm-1)

	best = f(0, 0) + build(0, 0, gn-1, gm-1)
	res = string(gpath[1:])
	return
}

func f(i int, j int) int {
	return (gx[i] + gy[j]) % gp
}

func dp(r1 int, c1 int, r2 int, c2 int, k int) []int {
	const neg = -1 << 60
	for i := r1; i <= r2; i++ {
		garr1[i] = neg
		garr2[i] = neg
	}
	garr1[r1] = f(r1, c1)
	prevLo, prevHi := r1, r1

	for x := r1 + c1 + 1; x <= k; x++ {
		lo := max(r1, x-c2)
		hi := min(r2, x-c1)

		for i := lo; i <= hi; i++ {
			j := x - i
			garr2[i] = f(i, j)
			bestHere := neg
			if i > r1 && prevLo <= i-1 && i-1 <= prevHi {
				bestHere = max(bestHere, garr1[i-1])
			}
			if j > c1 && prevLo <= i && i <= prevHi {
				bestHere = max(bestHere, garr1[i])
			}
			garr2[i] += bestHere
		}

		for i := prevLo; i <= prevHi; i++ {
			garr1[i] = neg
		}
		for i := lo; i <= hi; i++ {
			garr1[i] = garr2[i]
			garr2[i] = neg
		}
		prevLo, prevHi = lo, hi
	}
	return garr1
}

func fp(r1 int, c1 int, r2 int, c2 int, k int) []int {
	const neg = -1 << 60
	for i := r1; i <= r2; i++ {
		garr3[i] = neg
		garr2[i] = neg
	}
	garr3[r2] = f(r2, c2)
	prevLo, prevHi := r2, r2

	for x := r2 + c2 - 1; x >= k; x-- {
		lo := max(r1, x-c2)
		hi := min(r2, x-c1)
		for i := lo; i <= hi; i++ {
			j := x - i
			garr2[i] = f(i, j)
			bestHere := neg
			if i < r2 && prevLo <= i+1 && i+1 <= prevHi {
				bestHere = max(bestHere, garr3[i+1])
			}
			if j < c2 && prevLo <= i && i <= prevHi {
				bestHere = max(bestHere, garr3[i])
			}
			garr2[i] += bestHere
		}

		for i := prevLo; i <= prevHi; i++ {
			garr3[i] = neg
		}
		for i := lo; i <= hi; i++ {
			garr3[i] = garr2[i]
			garr2[i] = neg
		}
		prevLo, prevHi = lo, hi
	}
	return garr3
}

func build(r1 int, c1 int, r2 int, c2 int) int {
	var sum int
	if r1 == r2 {
		for j := c1; j < c2; j++ {
			gpath[r1+j+1] = 'S'
			sum += f(r1, j+1)
		}
		return sum
	}
	if c1 == c2 {
		for i := r1; i < r2; i++ {
			gpath[i+c1+1] = 'C'
			sum += f(i+1, c1)
		}
		return sum
	}

	k := (r1 + c1 + r2 + c2) / 2
	d1 := dp(r1, c1, r2, c2, k)
	d2 := fp(r1, c1, r2, c2, k)

	lo := max(r1, k-c2)
	hi := min(r2, k-c1)

	bestPair := []int{-1 << 60, r1}
	for i := lo; i <= hi; i++ {
		j := k - i
		tmp := d1[i] + d2[i] - f(i, j)
		if tmp > bestPair[0] {
			bestPair[0] = tmp
			bestPair[1] = i
		}
	}

	i := bestPair[1]
	j := k - bestPair[1]
	sum += build(r1, c1, i, j)
	sum += build(i, j, r2, c2)
	return sum
}

File: src/codeforces/set0/set1/set100/set101/e/cpp/solution.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, p;
    cin >> n >> m >> p;

    vector<int> x(n), y(m);
    for (int i = 0; i < n; ++i) {
        cin >> x[i];
    }
    for (int i = 0; i < m; ++i) {
        cin >> y[i];
    }

    vector<int> arr1(n), arr2(n), arr3(n);
    const int neg = -(1 << 30);
    auto f = [&](int i, int j) -> int {
        return (x[i] + y[j]) % p;
    };

    auto dp = [&](int r1, int c1, int r2, int c2, int k) -> vector<int>& {
        for (int i = r1; i <= r2; ++i) {
            arr1[i] = neg;
            arr2[i] = neg;
        }
        arr1[r1] = f(r1, c1);
        int prev_lo = r1;
        int prev_hi = r1;

        for (int diag = r1 + c1 + 1; diag <= k; ++diag) {
            int lo = max(r1, diag - c2);
            int hi = min(r2, diag - c1);
            for (int i = lo; i <= hi; ++i) {
                int j = diag - i;
                arr2[i] = f(i, j);
                int best_here = neg;
                if (i > r1 && prev_lo <= i - 1 && i - 1 <= prev_hi) {
                    best_here = max(best_here, arr1[i - 1]);
                }
                if (j > c1 && prev_lo <= i && i <= prev_hi) {
                    best_here = max(best_here, arr1[i]);
                }
                arr2[i] += best_here;
            }
            for (int i = prev_lo; i <= prev_hi; ++i) {
                arr1[i] = neg;
            }
            for (int i = lo; i <= hi; ++i) {
                arr1[i] = arr2[i];
                arr2[i] = neg;
            }
            prev_lo = lo;
            prev_hi = hi;
        }
        return arr1;
    };

    auto fp = [&](int r1, int c1, int r2, int c2, int k) -> vector<int>& {
        for (int i = r1; i <= r2; ++i) {
            arr3[i] = neg;
            arr2[i] = neg;
        }
        arr3[r2] = f(r2, c2);
        int prev_lo = r2;
        int prev_hi = r2;

        for (int diag = r2 + c2 - 1; diag >= k; --diag) {
            int lo = max(r1, diag - c2);
            int hi = min(r2, diag - c1);
            for (int i = lo; i <= hi; ++i) {
                int j = diag - i;
                arr2[i] = f(i, j);
                int best_here = neg;
                if (i < r2 && prev_lo <= i + 1 && i + 1 <= prev_hi) {
                    best_here = max(best_here, arr3[i + 1]);
                }
                if (j < c2 && prev_lo <= i && i <= prev_hi) {
                    best_here = max(best_here, arr3[i]);
                }
                arr2[i] += best_here;
            }
            for (int i = prev_lo; i <= prev_hi; ++i) {
                arr3[i] = neg;
            }
            for (int i = lo; i <= hi; ++i) {
                arr3[i] = arr2[i];
                arr2[i] = neg;
            }
            prev_lo = lo;
            prev_hi = hi;
        }
        return arr3;
    };

    string path(n + m - 1, '\0');

    auto build = [&](auto&& self, int r1, int c1, int r2, int c2) -> int {
        int sum = 0;
        if (r1 == r2) {
            for (int j = c1; j < c2; ++j) {
                path[r1 + j + 1] = 'S';
                sum += f(r1, j + 1);
            }
            return sum;
        }
        if (c1 == c2) {
            for (int i = r1; i < r2; ++i) {
                path[i + c1 + 1] = 'C';
                sum += f(i + 1, c1);
            }
            return sum;
        }

        int k = (r1 + c1 + r2 + c2) / 2;
        vector<int>& d1 = dp(r1, c1, r2, c2, k);
        vector<int>& d2 = fp(r1, c1, r2, c2, k);

        int lo = max(r1, k - c2);
        int hi = min(r2, k - c1);

        int best_score = neg;
        int best_row = r1;
        for (int i = lo; i <= hi; ++i) {
            int j = k - i;
            int cur = d1[i] + d2[i] - f(i, j);
            if (cur > best_score) {
                best_score = cur;
                best_row = i;
            }
        }

        int i = best_row;
        int j = k - best_row;
        sum += self(self, r1, c1, i, j);
        sum += self(self, i, j, r2, c2);
        return sum;
    };

    int best = f(0, 0) + build(build, 0, 0, n - 1, m - 1);
    cout << best << '\n';
    cout << path.substr(1) << '\n';
    return 0;
}
